### PR TITLE
system df: fix image-size calculations

### PR DESCRIPTION
**Pull request by vrothberg**
Fix two bugs in `system df`:

1.  The total size was calculated incorrectly as it was creating the sum of all image sizes but did not consider that a) the same image may be listed more than once (i.e., for each repo-tag pair), and that b) images share layers.

    The total size is now calculated directly in `libimage` by taking multi-layer use into account.

2.  The reclaimable size was calculated incorrectly.  This number indicates which data we can actually remove which means the total size minus what containers use (i.e., the "unique" size of the image in use by containers).

Fixes: #16135
Signed-off-by: Valentin Rothberg <vrothberg@redhat.com>

#### Does this PR introduce a user-facing change?

```release-note
Fix a bug where `podman system df` reported wrong image sizes.
```

### CODE REVIEW

File: test/system/320-system-df.bats [hunk 1]
```diff
@@ -52,10 +52,27 @@ function teardown() {
...
+
+    # Make sure that the return image "raw" size is correct
+    run_podman image inspect $IMAGE --format "{{.Size}}"
+    expectedSize="$output"
+
+    run_podman system df --format "{{.RawSize}}"
+    is "${lines[0]}" "$expectedSize" "raw image size is correct"
+
+    # Clean up and check reclaimable image data
+    run_podman system df --format '{{ .Reclaimable }}'
```
Comment (edsantiago) [inline test/system/320-system-df.bats:66]: nitty nit: inconsistent use of whitespace in formats, is that deliberate?

**Comment by vrothberg** [comment]
The failures seem to be triggered by containers/storage@main

**Review by edsantiago — CHANGES_REQUESTED** [review]
Could you run your new tests against `main`? They pass for me, which means you're not actually testing your changes.

File: test/system/320-system-df.bats [hunk 2]
```diff
@@ -52,10 +52,27 @@ function teardown() {
     is "${lines[10]}" "[0-9a-f]\{64\} *[01] * 0B" \
        "system df -v, 'Volumes', second line"
 
-    # Clean up
+    run_podman system df -v
+    run_podman system df
```
[the remark this answers is not in the document]
Comment (vrothberg) [reply test/system/320-system-df.bats:56]: Yes :) thanks!

**Comment by vrothberg** [comment]
> The failures seem to be triggered by containers/storage@main

For now I forced back the previous c/storage version via a `replace`.  @mtrmac @rhatdan PTAL

**Comment by vrothberg** [comment]
Easy reproducer among others:
```
$ go get github.com/containers/storage@main
$ make vendor && make podman
$ ./test/apiv2/test-apiv2 images
```

**Comment by mheon** [comment]
Code LGTM

**Review by edsantiago — CHANGES_REQUESTED** [review]
Tests are still broken, in the sense that they pass on `main`. You are not testing anything new. Can you please find a way to write tests that break on `main` and pass with this PR?

**Comment by vrothberg** [comment]
> Tests are still broken, in the sense that they pass on `main`. You are not testing anything new. Can you please find a way to write tests that break on `main` and pass with this PR?

Thanks for checking, Ed! Will do :+1: 

Also need to update the apiv2 tests:
```
[+0165s] not ok 993 [45-system] GET libpod/system/df : output
[+0165s] #  expected: {"Images":[],"Containers":[],"Volumes":[]}
[+0165s] #    actual: {"ImagesSize":0,"Images":[],"Containers":[],"Volumes":[]}
```

**Comment by vrothberg** [comment]
@edsantiago can you take another look at the tests?

**Comment by edsantiago** [comment]
ARGH! Sorry; left tests running this morning, moved on to something else. New test looks great, and fails (as it should) on main.  Thank you!
/lgtm

**Pull request merged**